impl/httpMetricsManager.go

```diff
@@ -67,20 +66,31 @@ func (m *HttpMetricsManager) GetThroughput() int {
 
 // return the sampletime(duration) at input percentile
 func (m *HttpMetricsManager) GetDurationAt(thre int) time.Duration {
-	if !sort.IsSorted(m.records) {
-		sort.Sort(m.records)
+	targetCount := thre * m.numRecords / 100
+	count := 0
+	keySet := []time.Duration{}
+	for key := range m.histogram {
+		keySet = append(keySet, key)
 	}
-	queryIndx := len(m.records) * thre / 100
-	return m.records[queryIndx].duration / time.Millisecond
+	sort.Slice(keySet, func(i, j int) bool { return keySet[i] < keySet[j] })
+	for _, key := range keySet {
+		count += m.histogram[key]
+		if count >= targetCount {
+			return key
+		}
+	}
+	return -1
 }
+
 func (m *HttpMetricsManager) GetMeanDuration() time.Duration {
-	sum := m.durationSum.sum
-	for i := m.durationSum.index + 1; i < len(m.records); i++ {
-		sum += m.records[i].duration
+	// Approach using histogram
+	sum := 0
+	numRecords := 0
+	for key := range m.histogram {
```

Comment:
Instead of looping and summing everytime on every call, why don't we keep the `sum` as an instance var?